File: src/conversation_search/core/git_utils.py
#!/usr/bin/env python3
"""Git utilities for resolving repository roots from project paths."""

import subprocess
from pathlib import Path
from typing import Optional


def resolve_repo_root(filesystem_path: str) -> Optional[str]:
    """
    Resolve the git repository root for a given filesystem path.

    For worktrees, resolves to the main repository root (not the worktree path).
    Returns None for non-git directories or invalid paths.
    """
    path = Path(filesystem_path).expanduser()
    if not path.exists():
        return None

    try:
        # First get the git common dir to handle worktrees
        result = subprocess.run(
            ["git", "-C", str(path), "rev-parse", "--git-common-dir"],
            capture_output=True, text=True, timeout=5
        )
        if result.returncode != 0:
            return None

        git_common_dir = result.stdout.strip()
        common_path = Path(git_common_dir)

        # If it's an absolute path, the repo root is its parent
        if common_path.is_absolute():
            # .git dir -> parent is repo root
            if common_path.name == ".git":
                return str(common_path.parent)
            return str(common_path.parent)

        # Relative path: resolve relative to the input path (CWD for git)
        resolved_common = (path / git_common_dir).resolve()

        # .git dir -> parent is repo root
        if resolved_common.name == ".git":
            return str(resolved_common.parent)
        return str(resolved_common.parent)

    except (subprocess.TimeoutExpired, OSError):
        return None

File: src/git_utils.rs
use std::path::PathBuf;
use std::process::Command;

/// Resolve the git repository root for a given filesystem path.
///
/// For worktrees, resolves to the main repository root (not the worktree path).
/// Returns None for non-git directories or invalid paths.
pub fn resolve_repo_root(filesystem_path: &str) -> Option<String> {
    let path = crate::db::expand_path(filesystem_path);
    if !path.exists() {
        return None;
    }

    let output = Command::new("git")
        .args(["-C", &path.to_string_lossy(), "rev-parse", "--git-common-dir"])
        .output()
        .ok()?;

    if !output.status.success() {
        return None;
    }

    let git_common_dir = String::from_utf8_lossy(&output.stdout).trim().to_string();
    let common_path = PathBuf::from(&git_common_dir);

    let resolved = if common_path.is_absolute() {
        common_path
    } else {
        // Relative path: resolve relative to the input path
        match (path.join(&git_common_dir)).canonicalize() {
            Ok(p) => p,
            Err(_) => return None,
        }
    };

    // .git dir -> parent is repo root
    if resolved.file_name().map_or(false, |n| n == ".git") {
        resolved.parent().map(|p| p.to_string_lossy().to_string())
    } else {
        Some(resolved.to_string_lossy().to_string())
    }
}
